### PR TITLE
Add Artifact Rejection (Rules, Evaluation, and Sequence Rejection)

### PR DESCRIPTION
**Pull request by tab-cmd**
- Added Evaluator (class) - which handles rules enabled based on parameters given
- Added Rule (class) - abstract base class that defines a rule to be tested against incoming data
- Added to parameters.json - toggle on/off for artifact rejection and individual rules, input fields for high and low voltage threshold values
- Implemented changes to copy_phrase.py to repeat sequence upon artifact detection and write broken rules to logger.
- Implemented changes to signal_model.py (CopyPhraseWrapper) to skip inference process upon artifact detection.

### CODE REVIEW

File: bcipy/signal/evaluate/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	rules (list of rule objects, defined in rules.py)
+	broken_rules (dict): {'rule name1': True/False (indicating violation), 'rule_name2': T/F, ... , 'rule_namen': T/F}"""
+
+	def __init__(self):
+
+		self.rules = []
+
+		self.broken_rules = {}
+
+		params_file = 'bcipy/parameters/parameters.json'
```
Comment (tab-cmd) [inline bcipy/signal/evaluate/evaluate.py]: get your parameters from caller

File: bcipy/parameters/parameters.json [hunk 2]
```diff
@@ -865,5 +865,45 @@
     "helpTip": "What should the name of the main log file be for each BciPy session",
     "recommended_values": "",
     "type": "str"
+  },
+  "Artifact Rejection": {
```
Comment (tab-cmd) [inline bcipy/parameters/parameters.json]: lowercase parameter values

File: bcipy/parameters/parameters.json [hunk 3]
```diff
@@ -865,5 +865,45 @@
...
+    "value": "true",
+    "section": "bci_config",
+    "readableName": "Artifact Rejection",
+    "helpTip": "Check data sequences for unwanted artifacts and rejects such sequences.",
+    "recommended_values": "",
+    "type": "bool"
+  },
+  "HighVoltage Threshold": {
+    "value": "true",
+    "section": "bci_config",
```
Comment (tab-cmd) [inline bcipy/parameters/parameters.json:879]: `artifact_rejection`

File: bcipy/signal/evaluate/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	rules, which raise flags if broken. Feeds warnings to artifact rejector
+	as suggestions. 
+	
+	Add rules given in parameters to evaluator's ruleset and set keys for broken_rules.
+	One heading per rule.
+
+	rules (list of rule objects, defined in rules.py)
+	broken_rules (dict): {'rule name1': True/False (indicating violation), 'rule_name2': T/F, ... , 'rule_namen': T/F}"""
+
+	def __init__(self):
```
Comment (tab-cmd) [inline bcipy/signal/evaluate/evaluate.py]: get necessary parameters for evaluator operation. parameters_dict, voltage_enabled, etc

File: bcipy/signal/evaluate/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+
+"""evaluate.py """
+
+
+from bcipy.signal.evaluate.rules import *
```
Comment (tab-cmd) [inline bcipy/signal/evaluate/evaluate.py]: never star import

File: bcipy/signal/evaluate/rules.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+"""rules.py"""
+
+from bcipy.helpers.load import load_json_parameters
+import numpy as np
+
+
+class Rule:
```
Comment (tab-cmd) [inline bcipy/signal/evaluate/rules.py]: ABC with defining @abstractmethods: https://docs.python.org/3/library/abc.html

File: bcipy/signal/evaluate/rules.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+			return True
+
+
+class LowVoltage(Rule):
+
+	"""Low Voltage Rule. Set low threshold for permitted voltage. """
+
+	def __init__(self):
+
+		Rule.__init__(self,"LowVoltage Value")
```
Comment (tab-cmd) [inline bcipy/signal/evaluate/rules.py]: self.name = "my custom name"

File: bcipy/parameters/parameters.json (hunk 2 above)
Comment (orbitalhybridization) [reply bcipy/parameters/parameters.json]: gotcha. I've lowercased all parameter values that I added!

File: bcipy/parameters/parameters.json (hunk 3 above)
Comment (orbitalhybridization) [reply bcipy/parameters/parameters.json:879]: gotcha. I also now have "highvoltage_threshold," "lowvoltage_threshold," etc.

File: bcipy/signal/evaluate/evaluate.py (hunk 5 above)
Comment (orbitalhybridization) [reply bcipy/signal/evaluate/evaluate.py]: amended! is now: from rules import HighVoltage, LowVoltage

File: bcipy/signal/evaluate/evaluate.py (hunk 4 above)
Comment (orbitalhybridization) [reply bcipy/signal/evaluate/evaluate.py]: now designed to get params from caller!

File: bcipy/signal/evaluate/rules.py (hunk 6 above)
Comment (orbitalhybridization) [reply bcipy/signal/evaluate/rules.py]: This is a cool module! Implemented Rule as an ABC and made "isBroken" an abstractmethod.

File: bcipy/signal/evaluate/rules.py (hunk 7 above)
Comment (orbitalhybridization) [reply bcipy/signal/evaluate/rules.py]: changed!

**Review by orbitalhybridization — COMMENTED** [review]
Addressed comments, and made subsequent changes!

**Pull request closed without merging**